ISPC multiple runtime libc dependencies on Windows

ISPC seems to depend on multiple libc versions for different things.

1. Doing `ispc --help` from the regular command line produces errors that `MSVCP140.dll` is missing, as well as `VCRUNTIME140.dll` and `VCRUNTIME140_1.dll`. Running the same command from Visual Studio command line removes those errors. I do not think those dependencies are necessary either way.

2. Compiling object file is impossible without libc, for example `ucrt`.

ISPC program compiled to sse2 and avx2:
```c
export void data_double(uniform int data[], uniform int data_count)
{
 foreach (i = 0 ... data_count)
 {
  data[i] *= 2;
 }
 assert(0);
}
```
C program:
```c
#pragma comment(linker, "/subsystem:console")
#pragma comment(linker, "/entry:enter")
#pragma comment(lib, "kernel32")
#pragma comment(lib, "ucrt") // this is necessary
#include "demo.ispc.h"

void enter(void)
{
 int data[20] = {0};
 for (int i = 0; i < 20; i++)
 {
  data[i] = i;
 }
 data_double(data, 20);
}
```
With `ucrt` removed, errors are produced:
```
demo.ispc.obj : error LNK2019: unresolved external symbol abort referenced in function __set_system_isa
demo.ispc_sse2.obj : error LNK2001: unresolved external symbol abort
demo.ispc_avx2.obj : error LNK2001: unresolved external symbol abort
demo.ispc_sse2.obj : error LNK2019: unresolved external symbol puts referenced in function data_double___un_3C_uni_3E_uni_sse2
demo.ispc_avx2.obj : error LNK2001: unresolved external symbol puts
```

`__set_system_isa` seems to correspond to dynamic dispatch of the program between architectures, the rest are due to `assert` being present.

In my opinion, libc is an unnecessary dependency, as printing procedures could be implemented without libc dependency trivially, and dynamic dispatch can use a Window intrinsic to get CPU features. I know that libc is basically necessary on Linux, as it acts as a platform layer of a kind, but on Windows the platform layer are all the standard dlls, such as `kernel32` and `user32` and the rest. `ucrt` is also present, but only since Windows 10, so using it would break compatibility with earlier versions.

Whatever is the decision of the team, I'm fine with it - whether you decide to stick with a bunch of different libcs, or use `ucrt`, or remove the dependency altogether. It's merely inconvenient that I will have to work around it by doing my own dispatch, since in my program I do not link to any libc, to avoid having a user environment dependency.

--- a/builtins/dispatch.c
+++ b/builtins/dispatch.c
@@ -21,9 +21,13 @@
 #error "Either REGULAR or MACOS macro need to defined"
 #endif
 
-void abort();
-
 static int __system_best_isa = -1;
+
+void __terminate_now() {
+    // Terminate execution using x86 UD2 instruction. UD2 raises an invalid
+    // opcode exception, ensuring program termination.
+    __asm__ __volatile__("ud2");
+}
 
 static void __cpuid(int info[4], int infoType) {
     __asm__ __volatile__("cpuid" : "=a"(info[0]), "=b"(info[1]), "=c"(info[2]), "=d"(info[3]) : "0"(infoType));
@@ -146,7 +150,7 @@ static int __get_system_isa() {
         if (skx) {
             return 7; // SKX
         } else if (knl) {
-            abort(); // KNL
+            __terminate_now(); // KNL
         }
         // If it's unknown AVX512 target, fall through and use AVX2
         // or whatever is available in the machine.
@@ -167,9 +171,12 @@ static int __get_system_isa() {
         return 1; // SSE4.1
     } else if (sse2) {
         return 0; // SSE2
-    } else {
-        abort();
     }
+
+    __terminate_now();
+
+    // This is actually unreachable, but needed to suppress the warning.
+    return -1;
 }
 
 void __set_system_isa() {

--- a/docs/ispc.rst
+++ b/docs/ispc.rst
@@ -693,6 +693,13 @@ deprecated. It leads to a warning when the function is called.
 
 The target ``avx512knl-x16`` was removed.
 
+The release introduces a breaking change by altering the behavior of user
+programs when no supported ISA is detected in the auto-dispatch code. It
+replaces the previous ``SIGABRT`` signal with ``SIGILL``. This affects users
+who rely on ``SIGABRT`` in their signal handlers for error handling or
+recovery. Such users must update their code to handle ``SIGILL`` instead. The
+change improves predictability and eliminates reliance of the dispatcher on the
+C standard library.
 
 Getting Started with ISPC
 =========================

--- a/src/builtins-decl.cpp
+++ b/src/builtins-decl.cpp
@@ -1,5 +1,5 @@
 /*
-  Copyright (c) 2024, Intel Corporation
+  Copyright (c) 2024-2025, Intel Corporation
 
   SPDX-License-Identifier: BSD-3-Clause
 */
@@ -349,6 +349,7 @@ DECL_BUILTIN_NAME(__task_index);
 DECL_BUILTIN_NAME(__task_index0);
 DECL_BUILTIN_NAME(__task_index1);
 DECL_BUILTIN_NAME(__task_index2);
+DECL_BUILTIN_NAME(__terminate_now);
 DECL_BUILTIN_NAME(__wasm_cmp_msk_eq);
 
 std::unordered_map<PersistentGroup, std::vector<const char *>> persistentGroups = {

--- a/src/builtins-decl.h
+++ b/src/builtins-decl.h
@@ -1,5 +1,5 @@
 /*
-  Copyright (c) 2024, Intel Corporation
+  Copyright (c) 2024-2025, Intel Corporation
 
   SPDX-License-Identifier: BSD-3-Clause
 */
@@ -397,6 +397,7 @@ extern const char *const __task_index;
 extern const char *const __task_index0;
 extern const char *const __task_index1;
 extern const char *const __task_index2;
+extern const char *const __terminate_now;
 extern const char *const __wasm_cmp_msk_eq;
 
 } // namespace builtin

--- a/src/builtins.cpp
+++ b/src/builtins.cpp
@@ -356,7 +356,7 @@ void ispc::LinkDispatcher(llvm::Module *module) {
     llvm::Module *dispatchBCModule = dispatch->getLLVMModule();
     lAddDeclarationsToModule(dispatchBCModule, module);
     lAddBitcodeToModule(dispatchBCModule, module);
-    llvm::StringSet<> dispatchFunctions = {builtin::__set_system_isa};
+    llvm::StringSet<> dispatchFunctions = {builtin::__set_system_isa, builtin::__terminate_now};
     lSetAsInternal(module, dispatchFunctions);
 }
 

--- a/src/module.cpp
+++ b/src/module.cpp
@@ -3469,12 +3469,13 @@ static void lCreateDispatchFunction(llvm::Module *module, llvm::Function *setISA
         bblock = nextBBlock;
     }
 
-    // We couldn't find a match that the current system was capable of
-    // running.  We'll call abort(); this is a bit of a blunt hammer--it
-    // might be preferable to call a user-supplied callback--ISPCError(...)
+    // We couldn't find a match that the current system was capable of running.
+    // We used to call abort() here but replaced it with call to our own
+    // implementation __terminate_now from builtins/dispatch.c
+    // It might be preferable to call a user-supplied callback--ISPCError(...)
     // or some such, but we don't want to start imposing too much of a
     // runtime library requirement either...
-    llvm::Function *abortFunc = module->getFunction("abort");
+    llvm::Function *abortFunc = module->getFunction(builtin::__terminate_now);
     Assert(abortFunc);
     llvm::CallInst::Create(abortFunc, "", bblock);
 

